KEP-4540: CPUManager strict-cpu-reservation policy option blog (#48371)
kubernetes/website: content/en/blog/_posts/2024-12-11-cpumanager-strict-cpu-reservation.md

Reviewer: nit: we write kubelet in all-lowercase

Author: yes, I checked and all occurrences of kubelet are already in lowercase.

---
layout: blog

> title: 'Kubernetes v1.32: New CPU Manager Static Policy Option: Strict CPU Reservation'

Reviewer: What do you think of:
```suggestion
title: 'Kubernetes v1.32 Adds A New CPU Manager Static Policy Option For Strict CPU Reservation'
```

draft: true
date: 2024-12-11
slug: cpumanager-strict-cpu-reservation
author: >
  [Jing Zhang](https://github.com/jingczhang) (Nokia)
---

> In Kubernetes v1.32, after years of community discussion, we are excited to introduce `strict-cpu-reservation` option for the [CPU Manager static policy](/docs/tasks/administer-cluster/cpu-management-policies/#static-policy-options). This feature is currently in alpha and hidden by default hence needs be enabled explicitly.

Reviewer: Try this:
```suggestion
In Kubernetes v1.32, after years of community discussion, we are excited to introduce a
`strict-cpu-reservation` option for the [CPU Manager static policy](/docs/tasks/administer-cluster/cpu-management-policies/#static-policy-options).
This feature is currently in alpha, with the associated policy hidden by default. You can only use the
policy if you explicitly enable the alpha behavior in your cluster.
```

## Understanding the feature

> The CPU Manager static policy is used to reduce latency or improve performance. The `reservedSystemCPUs` defines an explicit CPU set for OS system daemons and kubernetes system daemons. This option is specifically designed for Telco/NFV use cases where uncontrolled interrupts/timers may impact the workload performance. you can use this option to define the explicit cpuset for the system/kubernetes daemons as well as the interrupts/timers, so the rest CPUs on the system can be used exclusively for workloads, with less impact from uncontrolled interrupts/timers. More details of this parameter can be found on the [Explicitly Reserved CPU List](/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list) page.

Reviewer: `reservedSystemCPUs` isn't specifically designed for telco use cases. It's also used in other contexts that want similar outcomes.

For example, you could use this in an air defense system or a cluster helping with civil radar and air traffic control. I have no examples to share but you sure could.

Author: Actually, the above writing is a direct copy and paste from the existing writing:
https://kubernetes.io/docs/tasks/administer-cluster/reserve-compute-resources/#explicitly-reserved-cpu-list

Agreed with your comment, I will re-phrase.

If you want to protect your system daemons and interrupt processing, the obvious way is to use the `reservedSystemCPUs` option.

However, until now this isolation is only implemented for guaranteed pods with integer CPU requests. Admission only compares the CPU requests against the allocatable CPUs. Since the CPU limits are higher than the CPU requests, current implementation allows burstable and best-effort pods to use up the capacity of `reservedSystemCPUs` and causes host OS services to starve in real life deployments. This behavior also makes benchmarking infrastructure and workloads both inaccurate.

When this new `strict-cpu-reservation` policy option is enabled, the CPU Manager static policy will not allow any workload to use the reserved system CPU cores.

## Enabling the feature

To enable this feature, you need to turn on both the `CPUManagerPolicyAlphaOptions` feature gate and the `strict-cpu-reservation` policy option. And you need to remove the `/var/lib/kubelet/cpu_manager_state` file if it exists and restart kubelet.

> With the following Kubelet configuration:

Reviewer: ```suggestion
With the following kubelet configuration:
```

  CPUManagerPolicyOptions: true
  CPUManagerPolicyAlphaOptions: true
cpuManagerPolicy: static
cpuManagerPolicyOptions:
  strict-cpu-reservation: "true"
reservedSystemCPUs: "0,32,1,33,16,48"
...
```

When `strict-cpu-reservation` is not set or set to false:
```console
# cat /var/lib/kubelet/cpu_manager_state
{"policyName":"static","defaultCpuSet":"0-63","checksum":1058907510}
```

When `strict-cpu-reservation` is set to true:
```console
# cat /var/lib/kubelet/cpu_manager_state
{"policyName":"static","defaultCpuSet":"2-15,17-31,34-47,49-63","checksum":4141502832}
```

## Monitoring the feature

You can monitor the feature impact by checking the following CPU Manager counters:
- `cpu_manager_shared_pool_size_millicores`: report shared pool size, in millicores (e.g. 13500m)
- `cpu_manager_exclusive_cpu_allocation_count`: report exclusively allocated cores, counting full cores (e.g. 16)

Your best-effort workloads may starve if the `cpu_manager_shared_pool_size_millicores` count is zero for prolonged time.

We believe any system critical daemon like a log forwarder should not run as best-effort, but you can review and adjust the amount of CPU cores reserved as needed.

## Conclusion

> Strict CPU reservation is critical for Telco/NFV use cases. It is also a prerequisite for enabling the all-in-one type of deployments where workloads are placed on nodes serving combined master+worker+storage roles.

Reviewer: We don't use the term `master` any more; it's sensitive given the historical and current prevalence of slavery. Try “control plane“.

Author: taken.
